### PR TITLE
need to fix for ros-indigo-csm

### PR DESCRIPTION
**Pull request by k-okada**
- need to use pkg-config to find CSM library
- add depend to csm and csm depends on libgsl

### CODE REVIEW

**Comment by ubuntuslave** [comment]
@k-okada Thanks for the advice, Kei.

@130s Do we want to merge with @k-okada's commits?

**Comment by 130s** [comment]
+1

I think we just have to wait for the build issue of CSM package to be fixed, which also @k-okada initiated a work (https://github.com/AndreaCensi/csm/pull/13). If we merged this PR without having solved CSM build issue, I'm afraid users would start failing to build `laser_scan_matcher` package on their machines.

**Comment by k-okada** [comment]
csm build has done, please conform ( download them and check if this work with this PR) and merge & release
http://jenkins.ros.org/job/ros-indigo-csm_binarydeb_saucy_amd64/

File: laser_scan_matcher/CMakeLists.txt [hunk 1]
```diff
@@ -20,8 +16,13 @@ set( ROS_CXX_DEPENDENCIES
 find_package(catkin REQUIRED COMPONENTS ${ROS_CXX_DEPENDENCIES})
 find_package(PCL REQUIRED QUIET)
 
+#Include csm project code and compile it
```
Comment (130s) [inline laser_scan_matcher/CMakeLists.txt]: I think we're no longer compiling the code upon the building of laser_scan_matcher package.

**Comment by 130s** [comment]
@k-okada I opened https://github.com/k-okada/scan_tools/pull/1 to add a simple unit test and a trivial fix.

+1 again for this PR. I can run `demo.launch` with `csm` 1.0.2-2 installed via DEB.

---

Following is what I just noticed about how CSM is linked, and I don't know how important it is.

Currently in LSM, whether a csm library is available or not during runtime isn't checked AFAIK. And the test I proposed doesn't fail even if I remove `csm` before the test starts.
Then I found [CSM exports both dynamic and static libraries](https://github.com/AndreaCensi/csm/blob/6fe14f7d3ff0ea5961db0254b2c912aa11d1a783/sm/CMakeLists.txt#L192-L198). 

```
$ find /opt/ros/indigo -iname *csm*
:
/opt/ros/indigo/lib/libcsm-static.a
:
/opt/ros/indigo/lib/libcsm.so
```

When I only remove .so, `catkin build` passes. When only .a is removed, catkin build fails (see the detailed result below). I might be misunderstanding something, but looks like LSM always tries to link with a static one and that might be why the test doesn't fail without CSM library, as long as LSM is built successfully.
That said the unit test I suggested in https://github.com/k-okada/scan_tools/pull/1 may be less valuable.

```
$ rm -fr build/ devel/ install/
$ sudo apt-get install ros-indigo-csm
The following NEW packages will be installed:
  ros-indigo-csm
0 upgraded, 1 newly installed, 0 to remove and 1 not upgraded.
Need to get 0 B/297 kB of archives.
After this operation, 3,016 kB of additional disk space will be used.
Selecting previously unselected package ros-indigo-csm.
(Reading database ... 623480 files and directories currently installed.)
Preparing to unpack .../ros-indigo-csm_1.0.2-2trusty-20151108-0920-+0000_amd64.deb ...
Unpacking ros-indigo-csm (1.0.2-2trusty-20151108-0920-+0000) ...
Setting up ros-indigo-csm (1.0.2-2trusty-20151108-0920-+0000) ...
#All required rosdeps installed successfully
$ find /opt/ros/indigo -iname *csm*
/opt/ros/indigo/share/hrpsys/src/rtc/TorqueController/TwoDofControllerDynamicsModel.cpp
/opt/ros/indigo/share/hrpsys/src/rtc/TorqueController/TwoDofControllerDynamicsModel.h
/opt/ros/indigo/include/csm
/opt/ros/indigo/include/csm/csm.h
/opt/ros/indigo/include/csm/csm_all.h
/opt/ros/indigo/lib/libcsm-static.a
/opt/ros/indigo/lib/pkgconfig/csm.pc
/opt/ros/indigo/lib/libcsm.so
$ sudo rm /opt/ros/indigo/lib/libcsm.so
$ catkin build laser_scan_matcher Creating build space directory, '/home/rospasta/cws_nav/build' 
--------------------------------------------------------------------------------------------------------------
Profile:                     default
Extending:             [env] /home/rospasta/cws_sim/devel:/home/rospasta/cws_hironxo/install:/home/rospasta/cws_hironxo/devel:/opt/ros/indigo
Workspace:                   /home/rospasta/cws_nav
Source Space:       [exists] /home/rospasta/cws_nav/src
Build Space:        [exists] /home/rospasta/cws_nav/build
Devel Space:       [missing] /home/rospasta/cws_nav/devel
Install Space:     [missing] /home/rospasta/cws_nav/install
DESTDIR:                     None
--------------------------------------------------------------------------------------------------------------
Isolate Develspaces:         False
Install Packages:            True
Isolate Installs:            False
--------------------------------------------------------------------------------------------------------------
Additional CMake Args:       None
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
--------------------------------------------------------------------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
--------------------------------------------------------------------------------------------------------------
Workspace configuration appears valid.
-------------------------------------------------------------------------------------------------------------- 
Found '15' packages in 0.0 seconds. 
Starting ==> laser_scan_matcher                                                                               
Finished <== laser_scan_matcher [ 17.2 seconds ]                                                              
[build] Finished.                                                                                             
[build] Runtime: 17.3 seconds 
$ rm -fr build/ devel/ install/
$ sudo apt-get install --reinstall ros-indigo-csm
0 upgraded, 0 newly installed, 1 reinstalled, 0 to remove and 1 not upgraded.
Need to get 0 B/297 kB of archives.
After this operation, 0 B of additional disk space will be used.
(Reading database ... 623562 files and directories currently installed.)
Preparing to unpack .../ros-indigo-csm_1.0.2-2trusty-20151108-0920-+0000_amd64.deb ...
Unpacking ros-indigo-csm (1.0.2-2trusty-20151108-0920-+0000) over (1.0.2-2trusty-20151108-0920-+0000) ...
Setting up ros-indigo-csm (1.0.2-2trusty-20151108-0920-+0000) ...
$ find /opt/ros/indigo -iname *csm*
/opt/ros/indigo/share/hrpsys/src/rtc/TorqueController/TwoDofControllerDynamicsModel.cpp
/opt/ros/indigo/share/hrpsys/src/rtc/TorqueController/TwoDofControllerDynamicsModel.h
/opt/ros/indigo/include/csm
/opt/ros/indigo/include/csm/csm.h
/opt/ros/indigo/include/csm/csm_all.h
/opt/ros/indigo/lib/libcsm-static.a
/opt/ros/indigo/lib/pkgconfig/csm.pc
/opt/ros/indigo/lib/libcsm.so
$ sudo rm /opt/ros/indigo/lib/libcsm-static.a
$ catkin build laser_scan_matcher 
Creating build space directory, '/home/rospasta/cws_nav/build' 
--------------------------------------------------------------------------------------------------------------
Profile:                     default
Extending:             [env] /home/rospasta/cws_sim/devel:/home/rospasta/cws_hironxo/install:/home/rospasta/cws_hironxo/devel:/opt/ros/indigo
Workspace:                   /home/rospasta/cws_nav
Source Space:       [exists] /home/rospasta/cws_nav/src
Build Space:        [exists] /home/rospasta/cws_nav/build
Devel Space:       [missing] /home/rospasta/cws_nav/devel
Install Space:     [missing] /home/rospasta/cws_nav/install
DESTDIR:                     None
--------------------------------------------------------------------------------------------------------------
Isolate Develspaces:         False
Install Packages:            True
Isolate Installs:            False
--------------------------------------------------------------------------------------------------------------
Additional CMake Args:       None
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
--------------------------------------------------------------------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
--------------------------------------------------------------------------------------------------------------
Workspace configuration appears valid.
-------------------------------------------------------------------------------------------------------------- 
Found '15' packages in 0.0 seconds. 
Starting ==> laser_scan_matcher                                                                               

[laser_scan_matcher] ==> '/home/rospasta/cws_nav/build/laser_scan_matcher/build_env.sh /usr/bin/make --jobserver-fds=3,4 -j' in '/home/rospasta/cws_nav/build/laser_scan_matcher'
Scanning dependencies of target roscpp_generate_messages_cpp
Scanning dependencies of target nav_msgs_generate_messages_py
Scanning dependencies of target roscpp_generate_messages_py
Scanning dependencies of target tf_generate_messages_py
Scanning dependencies of target sensor_msgs_generate_messages_lisp
Scanning dependencies of target nodelet_generate_messages_py
[  0%] Scanning dependencies of target sensor_msgs_generate_messages_py
Scanning dependencies of target roscpp_generate_messages_lisp
Built target nav_msgs_generate_messages_py
[  0%] [  0%] [  0%] Built target roscpp_generate_messages_cpp
[  0%] [  0%] Built target roscpp_generate_messages_py
[  0%] Built target tf_generate_messages_py
Built target nodelet_generate_messages_py
Built target sensor_msgs_generate_messages_lisp
Built target sensor_msgs_generate_messages_py
[  0%] Scanning dependencies of target dynamic_reconfigure_generate_messages_cpp
Scanning dependencies of target rosgraph_msgs_generate_messages_py
Scanning dependencies of target tf_generate_messages_cpp
Scanning dependencies of target nodelet_generate_messages_lisp
Built target roscpp_generate_messages_lisp
Scanning dependencies of target nodelet_generate_messages_cpp
Scanning dependencies of target std_msgs_generate_messages_cpp
Scanning dependencies of target std_msgs_generate_messages_lisp
[  0%] [  0%] [  0%] [  0%] [  0%] [  0%] Built target rosgraph_msgs_generate_messages_py
Built target std_msgs_generate_messages_cpp
Built target dynamic_reconfigure_generate_messages_cpp
Built target nodelet_generate_messages_cpp
Built target tf_generate_messages_cpp
Built target std_msgs_generate_messages_lisp
[  0%] Built target nodelet_generate_messages_lisp
Scanning dependencies of target bond_generate_messages_lisp
Scanning dependencies of target bond_generate_messages_py
Scanning dependencies of target bond_generate_messages_cpp
Scanning dependencies of target rosgraph_msgs_generate_messages_cpp
Scanning dependencies of target actionlib_generate_messages_cpp
Scanning dependencies of target actionlib_generate_messages_lisp
Scanning dependencies of target geometry_msgs_generate_messages_lisp
Scanning dependencies of target dynamic_reconfigure_gencfg
[  0%] Built target bond_generate_messages_lisp
[  0%] [  0%] [  0%] [  0%] [  0%] Built target geometry_msgs_generate_messages_lisp
Built target bond_generate_messages_py
Built target bond_generate_messages_cpp
Built target actionlib_generate_messages_cpp
Built target rosgraph_msgs_generate_messages_cpp
[  0%] [  0%] Built target actionlib_generate_messages_lisp
Built target dynamic_reconfigure_gencfg
Scanning dependencies of target dynamic_reconfigure_generate_messages_py
Scanning dependencies of target dynamic_reconfigure_generate_messages_lisp
Scanning dependencies of target geometry_msgs_generate_messages_py
Scanning dependencies of target topic_tools_generate_messages_py
Scanning dependencies of target actionlib_generate_messages_py
Scanning dependencies of target nav_msgs_generate_messages_lisp
Scanning dependencies of target geometry_msgs_generate_messages_cpp
[  0%] [  0%] Scanning dependencies of target rosgraph_msgs_generate_messages_lisp
Built target dynamic_reconfigure_generate_messages_py
Built target dynamic_reconfigure_generate_messages_lisp
[  0%] [  0%] Built target geometry_msgs_generate_messages_py
[  0%] Built target topic_tools_generate_messages_py
[  0%] [  0%] [  0%] Built target geometry_msgs_generate_messages_cpp
Built target rosgraph_msgs_generate_messages_lisp
Scanning dependencies of target tf_generate_messages_lisp
Scanning dependencies of target sensor_msgs_generate_messages_cpp
Built target actionlib_generate_messages_py
Built target nav_msgs_generate_messages_lisp
Scanning dependencies of target std_msgs_generate_messages_py
[  0%] Scanning dependencies of target actionlib_msgs_generate_messages_cpp
[  0%] Scanning dependencies of target tf2_msgs_generate_messages_py
Scanning dependencies of target tf2_msgs_generate_messages_cpp
Scanning dependencies of target tf2_msgs_generate_messages_lisp
Built target sensor_msgs_generate_messages_cpp
Built target std_msgs_generate_messages_py
Scanning dependencies of target actionlib_msgs_generate_messages_lisp
[  0%] [  0%] [  0%] [  0%] Built target tf_generate_messages_lisp
Built target actionlib_msgs_generate_messages_cpp
Built target tf2_msgs_generate_messages_py
[  0%] Built target tf2_msgs_generate_messages_cpp
Built target tf2_msgs_generate_messages_lisp
[  0%] Built target actionlib_msgs_generate_messages_lisp
Scanning dependencies of target pcl_msgs_generate_messages_cpp
Scanning dependencies of target actionlib_msgs_generate_messages_py
Scanning dependencies of target pcl_msgs_generate_messages_lisp
Scanning dependencies of target pcl_msgs_generate_messages_py
Scanning dependencies of target nav_msgs_generate_messages_cpp
Scanning dependencies of target pcl_ros_gencfg
Scanning dependencies of target topic_tools_generate_messages_cpp
Scanning dependencies of target topic_tools_generate_messages_lisp
[  0%] [  0%] Built target pcl_msgs_generate_messages_cpp
Built target actionlib_msgs_generate_messages_py
[  0%] [  0%] [  0%] [  0%] [  0%] Built target topic_tools_generate_messages_lisp
Built target pcl_msgs_generate_messages_py
Built target pcl_msgs_generate_messages_lisp
[  0%] Built target pcl_ros_gencfg
Built target nav_msgs_generate_messages_cpp
Built target topic_tools_generate_messages_cpp
Scanning dependencies of target laser_scan_matcher
[ 33%] Building CXX object CMakeFiles/laser_scan_matcher.dir/src/laser_scan_matcher.cpp.o
Linking CXX shared library /home/rospasta/cws_nav/devel/lib/liblaser_scan_matcher.so
/usr/bin/ld: cannot find -lcsm-static
collect2: error: ld returned 1 exit status
make[2]: *** [/home/rospasta/cws_nav/devel/lib/liblaser_scan_matcher.so] Error 1
make[1]: *** [CMakeFiles/laser_scan_matcher.dir/all] Error 2
make: *** [all] Error 2
[laser_scan_matcher] <== '/home/rospasta/cws_nav/build/laser_scan_matcher/build_env.sh /usr/bin/make --jobserver-fds=3,4 -j' failed with return code '2'

Failed   <== laser_scan_matcher [ 10.0 seconds ]                                                              
[build] There were '1' errors:                                                                                

Failed to build package 'laser_scan_matcher' because the following command:

# Command to reproduce:
cd /home/rospasta/cws_nav/build/laser_scan_matcher && /home/rospasta/cws_nav/build/laser_scan_matcher/build_env.sh /usr/bin/make --jobserver-fds=3,4 -j; cd -

# Path to log:
cat /home/rospasta/cws_nav/build/build_logs/laser_scan_matcher.log

Exited with return code: 2 

[build] Runtime: 10.1 seconds 
```

**Comment by k-okada** [comment]
including isaac's test code and now it ready to merge,

as for static linke, I do not why they choose to link against static library but that behavior seems ok and main objective of the test is to check regression,so if someone changed csm and it did not generate same results, we can catch ( I'm not sure if master and csm_eigen produces same results, but this test may answer such question, so for me, it is very valuable)

**Comment by 130s** [comment]
I see.

@ubuntuslave We can merge this if you're okay.

**Comment by 130s** [comment]
travis config is now added. Triggering it by reopening.

**Comment by k-okada** [comment]
if you follow link few more times, you'll noticed that csm is not released
to public repo, only shadow-fixed.

## 

◉ Kei Okada

On Tue, Nov 10, 2015 at 10:27 AM, Isaac I.Y. Saito <notifications@github.com

**Comment by 130s** [comment]
> csm is not released to public repo, only shadow-fixed.

Looks like it. Handling it at https://github.com/ccny-ros-pkg/scan_tools/pull/33

**Comment by 130s** [comment]
travis passed. Tests were run. 
@k-okada How do the results look?

**Comment by 130s** [comment]
Ok, let me go ahead and merge this. Then I'll make a new DEB release right away.
Thank you @k-okada for putting all together!

**Pull request merged**